Volume rendering example using 3D texture

almarklein opened the pull request:
This example demonstrates volume rendering on float32 volumetric data. It is based on the work of #14530 and supports two different render styles, colormaps and setting contrast limits.

As part of this, this PR adds:
* A shader to do volume rendering using two different render sty;es (MIP and isosurface).
* Two colormap images (gray and Viridis).
* A medical volume image (of an implanted stent) in nrrd format.

There is room for improvement, some of which may be done in this PR:
* [x] the example could have some ui controls to select the render style etc. 
* [x] adjust aspect ratio when the window is resized
* [ ] fix origin so that slices can also be rendered (inside the volume)

A screenshot:
![afbeelding](https://user-images.githubusercontent.com/3015475/45034019-90ee1c00-b056-11e8-8308-11385608aea9.png)

almarklein committed:
Add 3D data and colormaps

almarklein committed:
Add volume rendering example

almarklein committed:
whitespace tweaks

almarklein commented:
cc @takahirox @artur-trzesiok  @izaid

almarklein committed:
fix volume render example aspect

Mugen87 commented:
Live link: https://rawgit.com/almarklein/three.js/volumeexample/examples/webgl_materials_texture3d_volume1.html

Can you please rebase so it's possible to use the live example with the latest dev build?

takahirox commented on examples/js/shaders/VolumeShader.js:
```diff
@@ -0,0 +1,320 @@
+// Author: Almar Klein
+// Description: Shaders to render 3D volumes using raycasting.
+// The applied techniques are based on similar implementations in the Visvis and Vispy projects.
+// This is not the only approach, therefore it's marked 1.
```
nit: We generally use this style

    /**
     * @author Almar Klein / url
     *
     * ....
     */

takahirox commented:
Thanks for the PR. Can you add this example into `examples/files.js`?

takahirox commented on examples/js/shaders/VolumeShader.js:
```diff
@@ -0,0 +1,320 @@
+// This is not the only approach, therefore it's marked 1.
+
+
+THREE.ShaderLib[ 'volumerender1' ] = {
```
We don't add shader code into `THREE.ShaderLib` from `examples/js/shaders`. Refer to the existing shader files in `examples/js/shaders/`. Perhaps `THREE.VolumeRender(ing)` would looks better?

takahirox replied:
I just realized Ocean example adds into `THREE.ShaderLib` now. But I think we generally don't from `examples`. Perhaps we need to update Ocean example...

almarklein replied:
done

almarklein replied:
Ha, then I may just have been unlucky, because that was indeed one of the examples that I looked at :)

almarklein committed:
volume render config via gui

almarklein committed:
address reviewer comments

almarklein commented:
> Can you add this example into examples/files.js?

Done

takahirox commented:
BTW are these assets license free? If so, how did you get? I tried to find license free 3D texture data but I couldn't make it.

fernandojsg commented:
> BTW are these assets license free? If so, how did you get? I tried to find license free 3D texture data but I couldn't make it.

It would be nice to add `stend.license` or `LICENSE.txt` on the folder explaining the origin of the file.

Also `Volume.js` and `VolumeSlice.js` are missing on the PR

takahirox commented:
`Volume.js` and `VolumeSlice.js` are already in `dev`.

fernandojsg commented:
> Volume.js and VolumeSlice.js are already in dev.

Ouch my fault O:)

gkjohnson commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,304 @@
+
+		function init() {
+
+			// The volume renderer does not work very well with perspective yet
```
Is this a limitation of the existing Volume.js code or a limitation of the newly added Volume Shader?

almarklein replied:
Good question. I am not entirely sure, to be honest. It could be a limitation of the shader technique, but it can also be due to subtle texture coordinate alignment stuff. I *think* that I recall having the same technique working with a perspective view in the VisPy project. I'd need to dive into this at some point.

almarklein replied:
I should add that the rendering does work with a perspective camera, but it gets "wobbly" at the edges at certain angles.

gkjohnson commented:
I know an orthographic camera is being used but it's a little odd that zooming doesn't cause the volume to get larger. Is it possible to center the rotation of the orbit controls about the middle of the volume and have zoom modify the ortho camera size, instead?

WestLangley commented:
Try `OrthographicTrackballControls` instead, and see if it works as you expect.

almarklein committed:
volume render example use orthograpic controls and clarify 3D data license

almarklein commented:
Changed to use `OrthographicTrackballControls`, much better :)

The stent data is a sample dataset from my PhD and has been placed in the public domain. (It's also a standard image in the [Imageio](https://github.com/imageio/imageio/) project.) I added a note about it being in the public domain in the comment section of the data (nrrd has an ASCI header).

Mugen87 commented:
Um, `NRRDLoader` currently throws the following error in the demo:

`Uncaught TypeError: Cannot read property 'split' of null`

almarklein committed:
fix nrrd file

almarklein commented:
Woops, I broke the file when editing its header. Fixed.

fernandojsg commented:
> The stent data is a sample dataset from my PhD and has been placed in the public domain. (It's also a standard image in the Imageio project.) I added a note about it being in the public domain in the comment section of the data (nrrd has an ASCI header).

Cool! Although I still feel that it could be nice to add the license text file on the folder as we usually do on other assets will be nice, as many people don't know that nrrd has an ascii header so they could know explicitly which license applies to it

almarklein committed:
add readme for nrrd model source/license

almarklein commented:
Fair enough :)  I added a readme.txt with some info about the data.

takahirox commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,305 @@
+				// THREEJS will select R32F (33326) based on the RedFormat and FloatType.
+				// Also see https://www.khronos.org/registry/webgl/specs/latest/2.0/#TEXTURE_TYPES_FORMATS_FROM_DOM_ELEMENTS_TABLE
+				// TODO: look the dtype up in the volume metadata
+				var texture = new THREE.Texture3D(volume.data, volume.xLength, volume.yLength, volume.zLength, THREE.RedFormat, THREE.FloatType);
```
We've updated `Texture3D` constructor (see #14861).
Could you update this line?

almarklein committed:
update use of Texture3D constructor

almarklein commented:
This is ready to go as far as I am concerned.

gkjohnson commented:
I'm still seeing that zooming is a bit broken -- I can zoom out but not back in. And rotating the model happens about the corner of the volume. Are you seeing that, as well? Other than that I think it's great to have a volume example.

almarklein committed:
center control rotation around center of volume

almarklein commented:
@gkjohnson Good point. I've fixed that now.

WestLangley commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -203,6 +203,7 @@
 			container.appendChild( renderer.domElement );
 
 			controls = new THREE.OrthographicTrackballControls( camera, renderer.domElement );
+			controls.target = new THREE.Vector3(64, 64, 128);
```
```javascript
controls.target.set( 64, 64, 128 );
```

almarklein replied:
@WestLangley thanks! Corrected. Still getting used to Three :)

almarklein committed:
correct previous change.

almarklein commented:
> I'm still seeing that zooming is a bit broken

I'm not seeing that, BTW

takahirox commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+<!DOCTYPE html>
```
nit: What does"`1`" of `webgl_materials_texture3d_volume1.html` indicate?

almarklein replied:
I was anticipating for an alternative volume example. But I see that other examples start with a non-numbered file and then add 2,3 etc. Renamed.

takahirox commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+				// Colormap textures
+				this.cmtextures = { viridis: new THREE.TextureLoader().load( 'textures/cm_viridis.png' ),
+									gray: new THREE.TextureLoader().load( 'textures/cm_gray.png' )
+								  }
```
Weird indications? And seems missing `;`.

almarklein replied:
fixed.

takahirox commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+
+				// Apply standard volume material uniform info
+				volmaterial.uniforms.u_data = { type: 't', value: texture };
+				volmaterial.uniforms.u_size = { type: 'v3', value: [volume.xLength, volume.yLength, volume.zLength] };
```
Kinda nit: You need extra spaces for brackets, like `func( 0, 1, 2 )` and `[ 0, 1, 2 ]`.
See the code style https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style%E2%84%A2

almarklein replied:
Good point, I addressed this throughout.

WestLangley commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+			//camera = new THREE.PerspectiveCamera( 60, window.innerWidth / window.innerHeight, 0.01, 1e10 );
+			camera = new THREE.OrthographicCamera()
+			camera.position.z = 1000;
+			this.camera.right = window.innerWidth / 5.0;
```
What are your world units?

almarklein replied:
I guess the world units are "voxels" now. IIRC one voxel is aout 0.5 mm.

WestLangley commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+				volmaterial.side = THREE.BackSide;  // The volume shader uses the backface as its "reference point"
+
+				// Apply standard volume material uniform info
+				volmaterial.uniforms.u_data = { type: 't', value: texture };
```
The `type` property should not be needed when specifying uniforms. Is it?

almarklein replied:
I hust tried; apparently not. Thanks!

takahirox commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+			renderer2 = new THREE.WebGLRenderer( {alpha : true} );
+			renderer2.setClearColor( 0x000000, 0 );
+			renderer2.setSize( insetWidth, insetHeight );
+			container2.appendChild( renderer2.domElement );
```
I haven't read the code in the detail yet but what's `renderer2` for?

almarklein replied:
Basically, I toke another example to have a starting point for this demo (I'm quite new to Three, as you might have guessed :)). I guess this is one of the oddities that came with it. All the xx2 stuff is to draw a small coordinate frame in the bottom right. I removed it, making the code less complex.

WestLangley commented:
@takahirox Thank you for your comments. I will have limited availability for awhile. I would be grateful if you could continue helping to improve this PR. :-)

takahirox commented:
@WestLangley Seems like we were occasionally reviewing at the same time hehe. Yeah I'll keep doing but I've been a bit busy lately too. I'm happy if anyone helps improving this PR.

almarklein committed:
fixed style

almarklein commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+
+			config = this.volumeConfig;
+
+			this.volmaterial.uniforms.u_clim = {type: 'v2', value: [config.clim1, config.clim2]};
```
Ah, I can set value directly. Learning as I go :)

almarklein commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+				// TODO: the texture coordinates currently map directly to world coordinates. That's why we translate
+				// the geometry above. But the extractSlice() below assume that the volume is centered at the origin.
+				// We'd need to add a material attribute with texture coordinates to fix this.
+				return;
```
Agreed

almarklein commented on examples/webgl_materials_texture3d_volume1.html:
```diff
@@ -0,0 +1,308 @@
+		}
+
+		function rotateAroundWorldAxis(object, axis, radians) {
+			var rotWorldMatrix = new THREE.Matrix4();
```
deleted

almarklein committed:
better way of using uniforms. Plus bit of cleanup.

almarklein committed:
remove inset and tweak control

almarklein commented:
> Personally, I would use OrbitControls, omit damping, and only render the scene when the camera (or gui) changes. 

I completely agree. If only we could tell it what side is up ... otherwise I'd consider reshaping my input data. I will dive into this for a bit.

> ... only render the scene when the camera changes

I found that I can do that with the orbit controls using `Icontrols.addEventListener( 'change', this.animate.bind(this) );`. Is it even possible for the trackbal controls?

WestLangley commented:
>Is it even possible for the trackball controls?

Not to my knowledge. Consider my suggestion above and switch to `OrbitControls` for now. You will need to call `render()` on window resizing and in the loader callback.

almarklein committed:
update stent data

almarklein committed:
Fix camera, draw only when needed, cleanup

almarklein commented:
@WestLangley @takahirox I've switched to the OrbitCamera, made the volume upright, and only draw when there's a change. Also did some more cleanup of the code. I'm quite happy about it now :)

WestLangley commented:
@almarklein Thanks for this PR!

Unfortunately, it seems you are not yet completely familiar with three.js, and I have identified 20+ changes that are required. To avoid a lot of back-and-forth, I recommend that this PR be merged, and I will file a separate PR to implement the necessary changes.

For your reference, see [this SO answer](https://stackoverflow.com/questions/17558085/three-js-orthographic-camera/17567292#17567292) for how to property instantiate an orthographic camera. The dimensions of the camera's frustum are in world units, and have nothing to do with the pixel resolution of the user's screen. I will fix this for you.

Also for your reference, loaders support an asynchronous `onLoad` callback from which you can call `render()`. You do not need to set timeouts or call `requestAnimationFrame` to force a redraw.

/ping @mrdoob This is OK to merge and I will file a follow-up PR.

mrdoob commented:
@WestLangley Sounds good!

mrdoob commented:
Thanks!
